File: flatbuffers/array/deserialize_sample154.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <iomanip>
#include "sample154_generated.h"

using namespace MySample::Sample154;

int main(int argc, char *argv[]) {
    if (argc <= 1) {
        std::cerr << "Error!!!" << std::endl;
        return 1;
    }

    std::string bin_filename = argv[1];
    std::cout << bin_filename << std::endl;

    std::ifstream file(bin_filename, std::ios::binary | std::ios::ate);
    if (!file) {
        std::cerr << "Error opening file!" << std::endl;
        return 1;
    }

    std::streamsize size = file.tellg();
    file.seekg(0, std::ios::beg);
    std::vector<uint8_t> buffer(size);
    if (!file.read(reinterpret_cast<char*>(buffer.data()), size)) {
        std::cerr << "Error reading file!" << std::endl;
        return 1;
    }

    std::cout << "Hex Representation: ";
    for (uint8_t byte : buffer) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(byte) << ' ';
    }
    std::cout << std::endl;

    auto verifier = flatbuffers::Verifier(buffer.data(), buffer.size());
    if (!verifier.VerifyBuffer<TopTable>()) {
        std::cerr << "Failed to verify buffer" << std::endl;
        return 1;
    }

    auto toptable = GetTopTable(buffer.data());
    auto struct1 = toptable->struct1();
    if (!struct1) {
        std::cerr << "Error deserializing FlatBuffers: Table1 is null" << std::endl;
        return 1;
    }

    // std::cout << struct1 << std::endl;

    try {
        const flatbuffers::Array<Struct2, 3> *s = struct1->struct2();
        std::cout << s[0].sint32() << std::endl;
        
    } catch (const std::exception &e) {
        std::cerr << "Error deserializing FlatBuffers: " << e.what() << std::endl;
        std::cerr << "Buffer size: " << buffer.size() << std::endl;
    }

    return 0;
}

File: flatbuffers/array/deserialize_sample154.py
import sys

from MySample.Sample154.TopTable import TopTable

def main():

    if len(sys.argv) <= 1:
        print('Error!!!')
        return

    bin_filename = sys.argv[1]
    print(sys.argv[1])

    with open(bin_filename, 'rb') as f:
        buf = f.read()
        buf = bytearray(buf)
    hex_representation = ' '.join(f'{byte:02x}' for byte in buf)
    print(hex_representation)

    toptable = TopTable.GetRootAsTopTable(buf, 0)
    struct1 = toptable.Struct1()
    print(struct1)
    try:
        length = struct1.Struct2Length()
        print(length)
    except Exception as e:
        print(f"Error deserializing FlatBuffers: {e}")
        print(f"Buffer size: {len(buf)}")

if __name__ == '__main__':
    main()


# 0c 00 00 00 00 00 06 00 10 00 04 00 06 00 00 00 01 00 00 00 02 00 00 00 03 00 00 00
# 0c 00 00 00 00 00 06 00 14 00 04 00 06 00 00 00 01 00 00 00 02 00 00 00 03 00 00 00 04 00 00 00
# 0c 00 00 00 00 00 06 00 08 00 04 00 06 00 00 00 80 00 00 00
# 0c 00 00 00 00 00 06 00 10 00 04 00 06 00 00 00 01 00 00 00 02 00 00 00 03 00 00 00


# 0c 00 00 00 00 00 06 00 10 00 04 00 06 00 00 00 01 00 00 00 02 00 00 00 03 00 00 00
# 0c 00 00 00 00 00 06 00 0a 00 04 00 06 00 00 00 0c 00 00 00 00 00 06 00 08 00 04 00 06 00 00 00 04 00 00 00 03 00 00 00 01 00 00 00 02 00 00 00 03 00 00 00